modules/internal/ChapelDomain.chpl

```diff
@@ -462,14 +460,35 @@ module ChapelDomain {
 
   // any combinations not handled by the above
 
-  inline operator ==(d1: domain, d2: domain) param {
-    return false;
+  private proc cmpError(d1, d2) {
+    if d1.isRectangular() then
+      compilerError("comparing a rectangular domain against",
+        " an associative or sparse domain is not currently supported");
+    if d1.isAssociative() then
+      compilerError("comparing an associative domain against",
+        " a rectangular or sparse domain is not currently supported");
+    if d1.isSparse() then
+      compilerError("comparing a sparse domain against",
+        " a rectangular or associative domain is not currently supported");
+    // we should not get here, however if we do...
+    compilerError("comparing '", d1.type:string, "' against '",
+                  d2.type:string, "' is not currently supported");
   }
 
-  inline operator !=(d1: domain, d2: domain) param {
-    return true;
+  inline operator ==(d1: domain, d2: domain) {
+    cmpError(d1, d2);
+  }
+
+  inline operator !=(d1: domain, d2: domain) {
+    cmpError(d1, d2);
   }
 
+  // check this before comparing two domains with == or !=
+  proc chpl_sameDomainKind(d1: domain, d2: domain) param
+    return (d1.isRectangular() && d2.isRectangular()) ||
```

Comment:
Is there no better way to do this than exhaustively comparing each of the domain types? We can't do something like: `d1.domainKind == d2.domainKind`

If we add a new domain type or change rectangular to mean (dense rectangular or sparse) we'd have to remember to update this code.

Reply:
Not that I know. I agree this implementation is not ideal. Also, it is tied to the current implementation of `==`. For example, if `==` allowed us to compare sparse against associative, this would return true for any combination of irregular domains. Maybe it should really be called "equalityIsImplemented". I updated the OP for this.